File: java/3092. Most Frequent IDs.java
class Solution {
    public long[] mostFrequentIDs(int[] nums, int[] freq) {
        int n = nums.length;
        long[] answer = new long[n];
        PriorityQueue<Ele> pq = new PriorityQueue<>((a,b) -> {
            if (b.count > a.count) return 1;
            else return -1;
        });
        Map<Integer, Long> count = new HashMap<>();
        for (int i = 0; i < n; i++) {
            count.put(nums[i], count.getOrDefault(nums[i], 0L) + freq[i]);
            pq.add(new Ele(nums[i], count.get(nums[i])));
            while (!pq.isEmpty() && (count.get(pq.peek().id) != pq.peek().count || count.get(pq.peek().id) == 0)) pq.poll();
            answer[i] = pq.isEmpty() ? 0 : pq.peek().count;
        }
        return answer;
    }
}

class Ele {
    int id;
    long count;
    Ele(int id, long count) {
        this.id = id;
        this.count = count;
    }
}

File: go/3092. Most Frequent IDs.go
func mostFrequentIDs(nums []int, freq []int) []int64 {
	n := len(nums)
	pq := make(Pq, 0)
	answer := make([]int64, 0)
	count := make(map[int]int)
	for i := 0; i < n; i++ {
		count[nums[i]] = count[nums[i]] + freq[i]
		heap.Push(&pq, NewEle(nums[i], count[nums[i]]))
		for len(pq) > 0 && (count[pq[0].id] != pq[0].count || count[pq[0].id] == 0) {
			heap.Pop(&pq)
		}
		if len(pq) == 0 {
			answer = append(answer, 0)
		} else {
			answer = append(answer, int64(pq[0].count))
		}
	}
	return answer
}

type Pq []Ele

func (pq Pq) Len() int {
	return len(pq)
}
func (pq Pq) Less(i, j int) bool {
	return pq[i].count > pq[j].count
}
func (pq Pq) Swap(i, j int) {
	pq[i], pq[j] = pq[j], pq[i]
}
func (pq *Pq) Push(x interface{}) {
	*pq = append(*pq, x.(Ele))
}
func (pq *Pq) Pop() interface{} {
	n := len(*pq)
	x := (*pq)[n-1]
	*pq = (*pq)[:n-1]
	return x
}

type Ele struct {
	id    int
	count int
}

func NewEle(id, count int) Ele {
	return Ele{
		id:    id,
		count: count,
	}
}